### DISCUSSION TITLE
Adding offset vector to mortar secondary/primary

### DISCUSSION THREAD

**Discussion by ttruster** [idaholab/moose#25802] [Q&A General] [answered]
Is there an interest to include the provision for an offset vector to translate the primary and secondary sides of a mortar surface? 

Currently, the search direction is only along the unit normal. For most applications that is fine; but in some cases the two blocks might be translated from each other. 

One potential use is for tensor mechanics periodic boundary value problems in which the RVE shape is not a rectangle. Then the paired surfaces still match up but are offset by some vector. For a rectangular RVE, the pairs are opposite along the normal direction.

I already implemented and verified it in 2d and 3d in the lower level functions such that `boundary_offset` is the only additional user parameter on the `MortarConstraint` objects, and its default value is a zero vector.

**Comment by lindsayad** [comment]
There's interest from at least one person ... you! Haha. IMO that's reason enough to consider putting it in the repo. I suppose the only issue would be if it really slowed mortar down ... but I suspect this is relatively inexpensive compared to a lot of the things we do with mortar. I don't personally have a use for this. I don't know if @recuero would. But anyway as I said if you would use it, then that's reason enough to consider it

**Reply by recuero** [reply]
I think that's a good feature to have in the code. I suspect it'd only add some logic if the offset option is selected?

**Reply by ttruster** [reply]
Right now I coded it as a `RealVectorValue` parameter that defaults to (0,0,0) if no value is provided. And that vector is then added to the nodal coordinates of the primary side in about 2 dozen places within about a dozen source files. 

The changes are shown in the last 3 commits of this branch, https://github.com/Truster-Research-Group/moose/tree/element-neighbor-pairing, which is obviously hugely behind 'next' right now.

**Reply by hugary1995** [reply]
I think this is a great feature to have! @ttruster have you profiled it? I doubt this will cause noticeable slowdown. But even if it does, we can always add a configure option because:
1. The user base for this feature is currently small.
2. Developers will know whether this feature is needed at the application level.

**Reply by ttruster** [reply, marked as answer]
Based on interest, I'll go ahead and open a pull request in the next couple days and continue discussion over there.
